diesel/src/query_builder/ast_pass.rs

```diff
@@ -50,25 +72,73 @@ impl<'a, DB> AstPass<'a, DB> where
         AstPass { internals }
     }
 
+    /// Mark the current query being constructed as unsafe to store in the
+    /// prepared statement cache.
+    ///
+    /// Diesel caches prepared statements as much as possible. However, it is
+    /// important to ensure that this doesn't result in unbounded memory usage
+    /// on the database server. To ensure this is the case, any logical query
+    /// which could generate a potentially unbounded number of prepared
+    /// statements *must* call this method. Examples of AST nodes which do this
+    /// are:
+    ///
+    /// - `SqlLiteral`. We have no way of knowing if the sql string was
+    ///   constructed dynamically or not, so we must assume it was dynamic.
+    /// - `EqAny` when passed a Rust `Vec`. The `IN` operator requires one bind
+    ///   parameter per element, meaning that the query could generate up to
+    ///   `usize` unique prepared statements.
+    /// - `InsertStatement`. Unbounded due to the variable number of records
+    ///   being inserted generating unique SQL.
+    /// - `UpdateStatement`. The number of potential queries is actually
+    ///   technically bounded, but the upper bound is the number of columns on
+    ///   the table factorial which is too large to be safe.
     pub fn unsafe_to_cache_prepared(&mut self) {
         if let AstPassInternals::IsSafeToCachePrepared(ref mut result) = self.internals {
             **result = false
         }
     }
 
+    /// Push the given SQL string on the end of the query being constructed.
+    ///
+    /// # Example
+    ///
+    /// ```ignore
+    /// impl<DB> QueryFragment<DB> for And<Left, Right>
+    /// where
+    ///     DB: Backend,
+    ///     Left: QueryFragment<DB>,
+    ///     Right: QueryFragment<DB>,
+    /// {
+    ///     pub fn walk_ast(&self, mut out: AstPass<DB>) -> QueryResult<()> {
+    ///         self.left.walk_ast(out.reborrow())?;
+    ///         out.push_sql(" AND ");
+    ///         self.right.walk_ast(out.reborrow())?;
+    ///         Ok(())
+    ///     }
+    /// }
+    /// ```
     pub fn push_sql(&mut self, sql: &str) {
         if let AstPassInternals::ToSql(ref mut builder) = self.internals {
             builder.push_sql(sql);
         }
     }
 
+    /// Push the given SQL identifier on the end of the query being constructed.
```

Comment:
> This would confuse me if I didn't know what it was doing.

That's semi-intentional. This is the method least likely to be used by plugins. I should perhaps call that out more explicitly